keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -589,6 +592,152 @@ associativeList.all(e, e.key1 == 'a' && e.key2 == 'b' && e.val == 100)
 associativeList.all(e, e.val == 100)
 ```
 
+### Resource constraints
+
+CEL expressions have the potential to consume unacceptable amounts of API server resources. We intend to constrain the resource utilization in two ways:
+
+- Proactively by static analysis of the CEL expressions to compute a "cost", and limiting that cost
+- Reactively by enforcing CEL expression evaluation time limits using go context cancelation
+
+We intend to enforce the limits both on a per-expression and per-request basis.
+
+Per-expression limits are primarily targeted at CEL expression authors and are intended to provide actionable feedback about expression cost. They are insufficient for ensuring that requests will be processed in a bounded amount of time. For example, a validation rule for a field within a list of objects, might be well within bounds each time it is evaluated, but the sum of the costs across the list of objects may be unacceptable. This is why we also have a per-request limit.
+
+#### Cost limits
+
+We will use CEL's [cost subsystem](https://github.com/google/cel-go/blob/dfef54b359b05532fb9695bc88937aa8530ab055/cel/program.go#L309) to provide our proactive limits. This is protects against both
+accidental and malicious misuse. If an expression is analyzed and has a cost greater than
+a predefined limit, it will not be allowed. If an expression is rejected for this reason, the error
+message will include how much the limit was exceeded.
+
+A major problem with the cost system is assigning the cost of list iteration. The cost of a CEL expression is 
+computed statically without any knowledge about the data that will be validated. Only the CEL expression is 
```

Comment:
Right. CEL is very constrained though. We're using a CEL runtime with a limit of 32 levels of expression nesting, 32 terms separated by `||` in a row, the expression can't define its own functions and has to fit with a CRD.  I'm curious what the worst case CEL expression would be here? It might still be pretty expensive? If so we might want to tighten some bounds. It seems containable to me, but I'd love to see someone prove me wrong (or right).